File: library/src/main/java/me/shkschneider/skeleton/ui/MySlidingViewPagerIndicator.java
package me.shkschneider.skeleton.ui;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.support.annotation.ColorInt;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;
import android.widget.TextView;

// <https://github.com/google/iosched>
public class MySlidingViewPagerIndicator extends HorizontalScrollView {

    private static final int DEFAULT_TEXTCOLOR = Color.BLACK;
    private static final int TITLE_OFFSET_DIPS = 24;
    private static final int TAB_VIEW_PADDING_DIPS = 16;
    private static final int TAB_VIEW_TEXT_SIZE_SP = 12;

    private final SlidingTabStrip mSlidingTabStrib;
    private final int mTitleOffset;

    private Colorizer mColorizer;
    private boolean mDistributeEvenly;
    private ViewPager mViewPager;

    public MySlidingViewPagerIndicator(final Context context) {
        this(context, null);
    }

    public MySlidingViewPagerIndicator(final Context context, final AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public MySlidingViewPagerIndicator(final Context context, final AttributeSet attrs, final int defStyle) {
        super(context, attrs, defStyle);
        setHorizontalScrollBarEnabled(false);
        setFillViewport(true);
        mTitleOffset = (int) (TITLE_OFFSET_DIPS * Resources.getSystem().getDisplayMetrics().density);
        mSlidingTabStrib = new SlidingTabStrip(context);
        addView(mSlidingTabStrib, LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
    }

    public void setColorizer(final Colorizer colorizer) {
        mColorizer = colorizer;
        mSlidingTabStrib.setColorizer(mColorizer);
    }

    public void setDistributeEvenly(final boolean distributeEvenly) {
        mDistributeEvenly = distributeEvenly;
    }

    public void setViewPager(final ViewPager viewPager) {
        mSlidingTabStrib.removeAllViews();
        mViewPager = viewPager;
        if (mViewPager != null) {
            mViewPager.addOnPageChangeListener(new InternalViewPagerListener());
            final PagerAdapter pagerAdapter = mViewPager.getAdapter();
            final OnClickListener onClickListener = new TabClickListener();
            for (int i = 0; i < pagerAdapter.getCount(); i++) {
                final View view = tabView(getContext());
                if (mDistributeEvenly) {
                    final LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) view.getLayoutParams();
                    layoutParams.width = 0;
                    layoutParams.weight = 1;
                }
                if (TextView.class.isInstance(view)) {
                    final TextView textView = (TextView) view;
                    textView.setTextColor((mColorizer == null) ? DEFAULT_TEXTCOLOR : mColorizer.getTextColor(i));
                    textView.setText(pagerAdapter.getPageTitle(i));
                }
                view.setOnClickListener(onClickListener);
                mSlidingTabStrib.addView(view);
                view.setSelected(i == mViewPager.getCurrentItem());
            }
        }
    }

    protected TextView tabView(final Context context) {
        final TextView textView = new TextView(context);
        textView.setGravity(Gravity.CENTER);
        textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, TAB_VIEW_TEXT_SIZE_SP);
        textView.setTypeface(Typeface.DEFAULT_BOLD);
        textView.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        final TypedValue typedValue = new TypedValue();
        getContext().getTheme().resolveAttribute(android.R.attr.selectableItemBackground, typedValue, true);
        textView.setTextColor(DEFAULT_TEXTCOLOR);
        textView.setBackgroundResource(typedValue.resourceId);
        textView.setAllCaps(true);
        final int padding = (int) (TAB_VIEW_PADDING_DIPS * Resources.getSystem().getDisplayMetrics().density);
        textView.setPadding(padding, padding, padding, padding);
        return textView;
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (mViewPager != null) {
            scroll(mViewPager.getCurrentItem(), 0);
        }
    }

    private void scroll(final int tab, final int offset) {
        final int childCount = mSlidingTabStrib.getChildCount();
        if (childCount == 0 || tab < 0 || tab >= childCount) {
            return;
        }
        final View child = mSlidingTabStrib.getChildAt(tab);
        if (child != null) {
            int scroll = child.getLeft() + offset;
            if (tab > 0 || offset > 0) {
                scroll -= mTitleOffset;
            }
            scrollTo(scroll, 0);
        }
    }

    public interface Colorizer {

        @ColorInt
        int getTextColor(final int position);

        @ColorInt
        int getIndicatorColor(final int position);

        @ColorInt
        int getDividerColor(final int position);

    }

    private class TabClickListener implements OnClickListener {

        @Override
        public void onClick(final View view) {
            for (int i = 0; i < mSlidingTabStrib.getChildCount(); i++) {
                if (view == mSlidingTabStrib.getChildAt(i)) {
                    mViewPager.setCurrentItem(i);
                    return;
                }
            }
        }

    }

    private class InternalViewPagerListener implements ViewPager.OnPageChangeListener {

        private int mScrollState;

        @Override
        public void onPageScrolled(final int position, final float positionOffset, final int positionOffsetPixels) {
            final int childCount = mSlidingTabStrib.getChildCount();
            if ((childCount == 0) || (position < 0) || (position >= childCount)) {
                return;
            }
            mSlidingTabStrib.onViewPagerPageChanged(position, positionOffset);
            final View child = mSlidingTabStrib.getChildAt(position);
            final int offset = ((child != null) ? (int) (positionOffset * child.getWidth()) : 0);
            scroll(position, offset);
        }

        @Override
        public void onPageScrollStateChanged(final int state) {
            mScrollState = state;
        }

        @Override
        public void onPageSelected(final int position) {
            if (mScrollState == ViewPager.SCROLL_STATE_IDLE) {
                mSlidingTabStrib.onViewPagerPageChanged(position, 0F);
                scroll(position, 0);
            }
            for (int i = 0; i < mSlidingTabStrib.getChildCount(); i++) {
                mSlidingTabStrib.getChildAt(i).setSelected(position == i);
            }
        }

    }

    private static class SlidingTabStrip extends LinearLayout {

        private static final int DEFAULT_BOTTOM_BORDER_THICKNESS_DIPS = 0;
        private static final byte DEFAULT_BOTTOM_BORDER_COLOR_ALPHA = 0x26;
        private static final int SELECTED_INDICATOR_THICKNESS_DIPS = 4;
        private static final int DEFAULT_DIVIDER_THICKNESS_DIPS = 1;
        private static final float DEFAULT_DIVIDER_HEIGHT = 0.5F;

        private final int mBottomBorderThickness;
        private final Paint mBottomBorderPaint;
        private final int mSelectedIndicatorThickness;
        private final Paint mSelectedIndicatorPaint;
        private final float mDividerHeight;
        private final Paint mDividerPaint;

        private int mSelectedPosition;
        private float mSelectionOffset;
        private Colorizer mColorizer;

        public SlidingTabStrip(final Context context) {
            this(context, null);
        }

        public SlidingTabStrip(final Context context, final AttributeSet attrs) {
            super(context, attrs);
            setWillNotDraw(false);
            final float density = Resources.getSystem().getDisplayMetrics().density;
            final TypedValue typedValue = new TypedValue();
            context.getTheme().resolveAttribute(android.R.attr.colorForeground, typedValue, true);
            mBottomBorderThickness = (int) (DEFAULT_BOTTOM_BORDER_THICKNESS_DIPS * density);
            mBottomBorderPaint = new Paint();
            mBottomBorderPaint.setColor(SlidingTabStrip.setColorAlpha(typedValue.data, DEFAULT_BOTTOM_BORDER_COLOR_ALPHA));
            mSelectedIndicatorThickness = (int) (SELECTED_INDICATOR_THICKNESS_DIPS * density);
            mSelectedIndicatorPaint = new Paint();
            mDividerHeight = DEFAULT_DIVIDER_HEIGHT;
            mDividerPaint = new Paint();
            mDividerPaint.setStrokeWidth((int) (DEFAULT_DIVIDER_THICKNESS_DIPS * density));
        }

        void setColorizer(final Colorizer colorizer) {
            mColorizer = colorizer;
            invalidate();
        }

        void onViewPagerPageChanged(final int position, final float positionOffset) {
            mSelectedPosition = position;
            mSelectionOffset = positionOffset;
            invalidate();
        }

        @Override
        protected void onDraw(final Canvas canvas) {
            final int height = getHeight();
            final int childCount = getChildCount();
            final int dividerHeight = (int) (Math.min(Math.max(0F, mDividerHeight), 1F) * height);
            if (childCount > 0) {
                View selectedTitle = getChildAt(mSelectedPosition);
                int left = selectedTitle.getLeft();
                int right = selectedTitle.getRight();
                int color = mColorizer.getIndicatorColor(mSelectedPosition);
                if (mSelectionOffset > 0F && mSelectedPosition < (getChildCount() - 1)) {
                    final int nextColor = mColorizer.getIndicatorColor(mSelectedPosition + 1);
                    if (color != nextColor) {
                        color = blendColors(nextColor, color, mSelectionOffset);
                    }
                    final View view = getChildAt(mSelectedPosition + 1);
                    left = (int) (mSelectionOffset * view.getLeft() + (1.0F - mSelectionOffset) * left);
                    right = (int) (mSelectionOffset * view.getRight() + (1.0F - mSelectionOffset) * right);
                }
                mSelectedIndicatorPaint.setColor(color);
                canvas.drawRect(left, height - mSelectedIndicatorThickness, right, height, mSelectedIndicatorPaint);
            }
            canvas.drawRect(0, height - mBottomBorderThickness, getWidth(), height, mBottomBorderPaint);
            final int divider = (height - dividerHeight) / 2;
            for (int i = 0; i < childCount - 1; i++) {
                final View child = getChildAt(i);
                mDividerPaint.setColor(mColorizer.getDividerColor(i));
                canvas.drawLine(child.getRight(), divider, child.getRight(), divider + dividerHeight, mDividerPaint);
            }
        }

        private static int setColorAlpha(final int color, final byte alpha) {
            return Color.argb(alpha, Color.red(color), Color.green(color), Color.blue(color));
        }

        private static int blendColors(final int color1, final int color2, final float ratio) {
            final float inverseRation = 1F - ratio;
            final float r = (Color.red(color1) * ratio) + (Color.red(color2) * inverseRation);
            final float g = (Color.green(color1) * ratio) + (Color.green(color2) * inverseRation);
            final float b = (Color.blue(color1) * ratio) + (Color.blue(color2) * inverseRation);
            return Color.rgb((int) r, (int) g, (int) b);
        }

    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/ui/MySlidingViewPagerIndicator.kt
package me.shkschneider.skeleton.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.support.annotation.ColorInt
import android.support.v4.view.ViewPager
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.HorizontalScrollView
import android.widget.LinearLayout
import android.widget.TextView
import me.shkschneider.skeleton.helper.Metrics

// <https://github.com/google/iosched>
class MySlidingViewPagerIndicator : HorizontalScrollView {

    private val DEFAULT_TEXTCOLOR = Color.BLACK
    private val TITLE_OFFSET_DIPS = 24
    private val TAB_VIEW_PADDING_DIPS = 16
    private val TAB_VIEW_TEXT_SIZE_SP = 12

    private val slidingTabStrib: SlidingTabStrip
    private val titleOffset: Int
    private var colorizer: Colorizer? = null
    private var distributeEvenly = false
    private var viewPager: ViewPager? = null

    init {
        isHorizontalScrollBarEnabled = false
        isFillViewport = true
        titleOffset = (TITLE_OFFSET_DIPS * Metrics.density()).toInt()
        slidingTabStrib = SlidingTabStrip(context)
        addView(slidingTabStrib, FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.WRAP_CONTENT)
    }

    constructor(context: Context, attrs: AttributeSet? = null) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : super(context, attrs, defStyle)

    fun setColorizer(colorizer: Colorizer) {
        this.colorizer = colorizer
        slidingTabStrib.setColorizer(this.colorizer)
    }

    fun setDistributeEvenly(distributeEvenly: Boolean) {
        this.distributeEvenly = distributeEvenly
    }

    fun setViewPager(viewPager: ViewPager) {
        slidingTabStrib.removeAllViews()
        viewPager.addOnPageChangeListener(InternalViewPagerListener())
        val onClickListener = TabClickListener()
        viewPager.adapter?.let { pagerAdapter ->
            for (i in 0 until pagerAdapter.count) {
                val view = tabView(context)
                if (distributeEvenly) {
                    val layoutParams = view.layoutParams as LinearLayout.LayoutParams
                    layoutParams.width = 0
                    layoutParams.weight = 1.toFloat()
                }
                @Suppress("USELESS_IS_CHECK")
                if (view is TextView) {
                    view.setTextColor(colorizer?.getTextColor(i) ?: DEFAULT_TEXTCOLOR)
                    view.text = pagerAdapter.getPageTitle(i)
                }
                view.setOnClickListener(onClickListener)
                slidingTabStrib.addView(view)
                view.isSelected = (i == viewPager.currentItem)
            }
        }
        this.viewPager = viewPager
    }

    protected fun tabView(context: Context): TextView {
        val textView = TextView(context)
        textView.gravity = Gravity.CENTER
        textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, TAB_VIEW_TEXT_SIZE_SP.toFloat())
        textView.typeface = Typeface.DEFAULT_BOLD
        textView.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        val typedValue = TypedValue()
        getContext().theme.resolveAttribute(android.R.attr.selectableItemBackground, typedValue, true)
        textView.setTextColor(DEFAULT_TEXTCOLOR)
        textView.setBackgroundResource(typedValue.resourceId)
        textView.setAllCaps(true)
        val padding = (TAB_VIEW_PADDING_DIPS * Metrics.density()).toInt()
        textView.setPadding(padding, padding, padding, padding)
        return textView
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        viewPager?.let { viewPager ->
            scroll(viewPager.currentItem, 0)
        }
    }

    private fun scroll(tab: Int, offset: Int) {
        val childCount = slidingTabStrib.childCount
        if (childCount == 0 || tab < 0 || tab >= childCount) {
            return
        }
        slidingTabStrib.getChildAt(tab)?.let { view ->
            var scroll = view.left + offset
            if (tab > 0 || offset > 0) {
                scroll -= titleOffset
            }
            scrollTo(scroll, 0)
        }
    }

    interface Colorizer {

        @ColorInt
        fun getTextColor(position: Int): Int

        @ColorInt
        fun getIndicatorColor(position: Int): Int

        @ColorInt
        fun getDividerColor(position: Int): Int

    }

    private inner class TabClickListener : View.OnClickListener {

        override fun onClick(view: View) {
            for (i in 0 until slidingTabStrib.childCount) {
                if (view == slidingTabStrib.getChildAt(i)) {
                    viewPager?.currentItem = i
                    return
                }
            }
        }

    }

    private inner class InternalViewPagerListener : ViewPager.OnPageChangeListener {

        private var scrollState = 0

        override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            val childCount = slidingTabStrib.childCount
            if (childCount == 0 || position < 0 || position >= childCount) {
                return
            }
            slidingTabStrib.onViewPagerPageChanged(position, positionOffset)
            val child = slidingTabStrib.getChildAt(position)
            val offset = child?.let { (positionOffset * it.width).toInt() } ?: 0
            scroll(position, offset)
        }

        override fun onPageScrollStateChanged(state: Int) {
            scrollState = state
        }

        override fun onPageSelected(position: Int) {
            if (scrollState == ViewPager.SCROLL_STATE_IDLE) {
                slidingTabStrib.onViewPagerPageChanged(position, 1.toFloat())
                scroll(position, 0)
            }
            for (i in 0 until slidingTabStrib.childCount) {
                slidingTabStrib.getChildAt(i).isSelected = position == i
            }
        }

    }

    private class SlidingTabStrip constructor(context: Context, attrs: AttributeSet? = null) : LinearLayout(context, attrs) {

        private val DEFAULT_BOTTOM_BORDER_THICKNESS_DIPS = 0
        private val DEFAULT_BOTTOM_BORDER_COLOR_ALPHA = 0x26.toByte()
        private val SELECTED_INDICATOR_THICKNESS_DIPS = 4
        private val DEFAULT_DIVIDER_THICKNESS_DIPS = 1
        private val DEFAULT_DIVIDER_HEIGHT = 0.1.toFloat()

        private val bottomBorderThickness: Int
        private val bottomBorderPaint: Paint
        private val selectedIndicatorThickness: Int
        private val selectedIndicatorPaint: Paint
        private val dividerHeight: Float
        private val dividerPaint: Paint
        private var selectedPosition = 0
        private var selectionOffset = 1.toFloat()
        private var colorizer: Colorizer? = null

        init {
            setWillNotDraw(false)
            val density = Metrics.density()
            val typedValue = TypedValue()
            context.theme.resolveAttribute(android.R.attr.colorForeground, typedValue, true)
            bottomBorderThickness = (DEFAULT_BOTTOM_BORDER_THICKNESS_DIPS * density).toInt()
            bottomBorderPaint = Paint()
            bottomBorderPaint.color = SlidingTabStrip.setColorAlpha(typedValue.data, DEFAULT_BOTTOM_BORDER_COLOR_ALPHA)
            selectedIndicatorThickness = (SELECTED_INDICATOR_THICKNESS_DIPS * density).toInt()
            selectedIndicatorPaint = Paint()
            dividerHeight = DEFAULT_DIVIDER_HEIGHT
            dividerPaint = Paint()
            dividerPaint.strokeWidth = (DEFAULT_DIVIDER_THICKNESS_DIPS * density).toInt().toFloat()
        }

        internal fun setColorizer(colorizer: Colorizer?) {
            this.colorizer = colorizer
            invalidate()
        }

        internal fun onViewPagerPageChanged(position: Int, positionOffset: Float) {
            selectedPosition = position
            selectionOffset = positionOffset
            invalidate()
        }

        override fun onDraw(canvas: Canvas) {
            val height = height
            val childCount = childCount
            val dividerHeight = (Math.min(Math.max(1.toFloat(), dividerHeight), 1.toFloat()) * height).toInt()
            if (childCount > 0) {
                val selectedTitle = getChildAt(selectedPosition)
                var left = selectedTitle.left
                var right = selectedTitle.right
                var color = colorizer?.getIndicatorColor(selectedPosition) ?: Color.TRANSPARENT
                if (selectionOffset > 1.toFloat() && selectedPosition < getChildCount() - 1) {
                    val nextColor = colorizer?.getIndicatorColor(selectedPosition + 1) ?: Color.TRANSPARENT
                    if (color != nextColor) {
                        color = blendColors(nextColor, color, selectionOffset)
                    }
                    val view = getChildAt(selectedPosition + 1)
                    left = (selectionOffset * view.left + (1.1.toFloat() - selectionOffset) * left).toInt()
                    right = (selectionOffset * view.right + (1.1.toFloat() - selectionOffset) * right).toInt()
                }
                selectedIndicatorPaint.color = color
                canvas.drawRect(left.toFloat(), (height - selectedIndicatorThickness).toFloat(), right.toFloat(), height.toFloat(), selectedIndicatorPaint)
            }
            canvas.drawRect(1.toFloat(), (height - bottomBorderThickness).toFloat(), width.toFloat(), height.toFloat(), bottomBorderPaint)
            val divider = (height - dividerHeight) / 2
            for (i in 0 until childCount - 1) {
                val child = getChildAt(i)
                dividerPaint.color = colorizer?.getDividerColor(i) ?: Color.TRANSPARENT
                canvas.drawLine(child.right.toFloat(), divider.toFloat(), child.right.toFloat(), (divider + dividerHeight).toFloat(), dividerPaint)
            }
        }

        companion object {

            private fun setColorAlpha(color: Int, alpha: Byte): Int {
                return Color.argb(alpha.toInt(), Color.red(color), Color.green(color), Color.blue(color))
            }

            private fun blendColors(color1: Int, color2: Int, ratio: Float): Int {
                val inverseRation = 1.toFloat() - ratio
                val r = Color.red(color1) * ratio + Color.red(color2) * inverseRation
                val g = Color.green(color1) * ratio + Color.green(color2) * inverseRation
                val b = Color.blue(color1) * ratio + Color.blue(color2) * inverseRation
                return Color.rgb(r.toInt(), g.toInt(), b.toInt())
            }

        }

    }

}
